src/hmem_cuda.c

```diff
@@ -484,29 +492,46 @@ static int cuda_hmem_verify_devices(void)
 	 * call to cudaGetDeviceCount() when possible.
 	 */
 
-	/* Make certain that the NVML routines are initialized */
-	nvml_ret = ofi_nvmlInit_v2();
-	if (nvml_ret != NVML_SUCCESS)
-                return -FI_ENOSYS;
+	/* If we are dlopen'ing the CUDA libraries, allow for the case
+	 * where the CUDA libraries are present, but the NVML library is
+	 * not present.  Set nvml_device_count to 1 in that case, bypassing
+	 * actual device count check.
+	 */
+#if ENABLE_CUDA_DLOPEN
+	if (!cuda_attr.nvml_handle)
+		nvml_device_count = 1;
+#endif
 
-	/* Verify NVIDIA devices are present on the host. */
-	nvml_ret = ofi_nvmlDeviceGetCount_v2(&nvml_device_count);
-	if (nvml_ret != NVML_SUCCESS) {
-	        ofi_nvmlShutdown();
-                return -FI_ENOSYS;
-        }
+	/* Check for NVIDIA devices, if NVML library is dlopen-ed*/
+	if (nvml_device_count == 0) {
 
-	/* Make certain that the NVML routines get shutdown */
-        /* Note: nvmlInit / Shutdown calls are refcounted, so no harm in
-         * calling nvmlShutdown here, if the user has called nvmlInit.
-         */
-	nvml_ret = ofi_nvmlShutdown();
-	if (nvml_ret != NVML_SUCCESS)
-                return -FI_ENOSYS;
+		/* Make certain that the NVML routines are initialized */
+		nvml_ret = ofi_nvmlInit_v2();
+		if (nvml_ret != NVML_SUCCESS)
+			return -FI_ENOSYS;
+
+		/* Verify NVIDIA devices are present on the host. */
+		nvml_ret = ofi_nvmlDeviceGetCount_v2(&nvml_device_count);
+		if (nvml_ret != NVML_SUCCESS) {
+			ofi_nvmlShutdown();
+			return -FI_ENOSYS;
+		}
 
-        FI_INFO(&core_prov, FI_LOG_CORE,
-                "Number of NVIDIA devices detected: %u\n",
-                nvml_device_count);
+		/* Make certain that the NVML routines get shutdown */
+		/* Note: nvmlInit / Shutdown calls are refcounted, so no harm in
+		 * calling nvmlShutdown here, if the user has called nvmlInit.
+		 */
+		nvml_ret = ofi_nvmlShutdown();
+		if (nvml_ret != NVML_SUCCESS)
+			return -FI_ENOSYS;
+
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Number of NVIDIA devices detected: %u\n",
+			nvml_device_count);
+	} else {
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"Skipping check for NVIDIA devices with NVML routines\n");
+	}
 
         /* If NVIDIA devices are present, now perform more expensive check
          * for actual GPUs.
```

Comment:
The check below here would be something like ```if (!nvml_handle || nvml_device_count)```.  Though we probably don't even need checks here.  If an nvml_device wasn't found, we should have already returned from the call.

Reply:
We need the check, for the case when the NVML library failed to dlopen() and for when it was opened and found devices that we are confirming with cudaGetDeviceCount.